src/aws_encryption_sdk/keyrings/aws_kms/__init__.py

```diff
@@ -0,0 +1,376 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Keyring for use with AWS Key Management Service (KMS).
+
+.. versionadded:: 1.5.0
+
+"""
+import logging
+
+import attr
+import six
+from attr.validators import deep_iterable, instance_of, optional
+
+from aws_encryption_sdk.exceptions import DecryptKeyError, EncryptKeyError
+from aws_encryption_sdk.identifiers import AlgorithmSuite
+from aws_encryption_sdk.internal.validators import value_is_not_a_string
+from aws_encryption_sdk.keyrings.base import Keyring
+from aws_encryption_sdk.keyrings.multi import MultiKeyring
+from aws_encryption_sdk.materials_managers import DecryptionMaterials, EncryptionMaterials
+from aws_encryption_sdk.structures import EncryptedDataKey, KeyringTrace, KeyringTraceFlag, MasterKeyInfo, RawDataKey
+
+from .client_suppliers import ClientSupplier, DefaultClientSupplier
+
+try:  # Python 3.5.0 and 3.5.1 have incompatible typing modules
+    from typing import Any, Dict, Iterable, Union  # noqa pylint: disable=unused-import
+except ImportError:  # pragma: no cover
+    # We only actually need these imports when running the mypy checks
+    pass
+
+__all__ = ("KmsKeyring",)
+
+_LOGGER = logging.getLogger(__name__)
+_PROVIDER_ID = "aws-kms"
+_GENERATE_FLAGS = {KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY}
+_ENCRYPT_FLAGS = {KeyringTraceFlag.WRAPPING_KEY_ENCRYPTED_DATA_KEY, KeyringTraceFlag.WRAPPING_KEY_SIGNED_ENC_CTX}
+_DECRYPT_FLAGS = {KeyringTraceFlag.WRAPPING_KEY_DECRYPTED_DATA_KEY, KeyringTraceFlag.WRAPPING_KEY_VERIFIED_ENC_CTX}
+
+
+@attr.s
+class KmsKeyring(Keyring):
+    """Keyring that uses AWS Key Management Service (KMS) Customer Master Keys (CMKs) to manage wrapping keys.
+
+    Set ``generator_key_id`` to require that the keyring use that CMK to generate the data key.
+    If you do not set ``generator_key_id``, the keyring will not generate a data key.
+
+    Set ``child_key_ids`` to specify additional CMKs that the keyring will use to encrypt the data key.
+
+    The keyring will attempt to use any CMKs
+    identified by CMK ARN in either ``generator_key_id`` or ``child_key_ids`` on decrypt.
+
+    You can identify CMKs by any `valid key ID`_ for the keyring to use on encrypt,
+    but for the keyring to attempt to use them on decrypt
+    you MUST specify the CMK ARN.
+
+    If you specify neither ``generator_key_id`` nor ``child_key_ids``
```

Comment:
When talking about this in the ESDK-JS we went with "explicit is better"
so you must pass `discovery : true`.
I'm not sure of the best way to do this in Python,
or if this is crazy.

Reply:
For now I'm ok with a derived property.